src/core/ext/transport/chttp2/transport/hpack_encoder.cc

```diff
@@ -54,27 +60,182 @@
 /* if the probability of this item being seen again is < 1/x then don't add
    it to the table */
 #define ONE_ON_ADD_PROBABILITY (GRPC_CHTTP2_HPACKC_NUM_VALUES >> 1)
-/* don't consider adding anything bigger than this to the hpack table */
-#define MAX_DECODER_SPACE_USAGE 512
-
-#define DATA_FRAME_HEADER_SIZE 9
+/* Meaningful to encoder and parser on remote end. */
+typedef uint32_t HpackEncoderIndex;
+/* Internal-table bookkeeping (*not* the hpack index). */
+typedef uint32_t HpackEncoderSlotHash;
+
+struct SliceRefComparator {
+  typedef grpc_slice_refcount* Type;
+  static grpc_slice_refcount* Null() { return nullptr; }
+  static bool Equals(grpc_slice_refcount* s1, grpc_slice_refcount* s2) {
+    return s1 == s2;
+  }
+  static void Ref(grpc_slice_refcount* sref) { sref->Ref(); }
+  static void Unref(grpc_slice_refcount* sref) { sref->Unref(); }
+};
 
-static grpc_slice_refcount terminal_slice_refcount(
-    grpc_slice_refcount::Type::STATIC);
-static const grpc_slice terminal_slice = {
-    &terminal_slice_refcount, /* refcount */
-    {{0, nullptr}}            /* data.refcounted */
+struct MetadataComparator {
+  typedef grpc_mdelem Type;
+  static const grpc_mdelem Null() { return {0}; }
+  static bool Equals(const grpc_mdelem md1, const grpc_mdelem md2) {
+    return md1.payload == md2.payload;
+  }
+  static void Ref(grpc_mdelem md) { GRPC_MDELEM_REF(md); }
+  static void Unref(grpc_mdelem md) { GRPC_MDELEM_UNREF(md); }
 };
 
+/* Index table management */
+template <typename H>
+static HpackEncoderIndex HpackIndex(const H* hashtable,
+                                    HpackEncoderSlotHash hash_index) {
+  return hashtable[hash_index].index;
+}
+
+template <typename R, typename H>
+static const R& GetEntry(const H* hashtable, HpackEncoderSlotHash hash_index) {
+  return hashtable[hash_index].value;
+}
+
+template <typename Cmp, typename H>
+static bool TableEmptyAt(const H* hashtable, HpackEncoderSlotHash hash_index) {
+  return Cmp::Equals(hashtable[hash_index].value, Cmp::Null());
+}
+
+template <typename Cmp, typename H, typename T>
```

Comment:
Hypothetically you could make a template class, that implements a hash table, that takes as template parameter the value type - which is what I did first. However, it does complicate the of hpack_encoder.h (specifically the definition of grpc_chttp2_hpack_compressor).

If we end up converting that class to C++, then it may make sense to have this be an HpackEncoderTable, but I didn't want to have a mix of C-style and C++ style definitions in the meantime.